### DISCUSSION TITLE
How to get ClassName from TypeElement with kotlinpoet-metadata

### DISCUSSION THREAD

**Discussion by GreyTeardrop** [square/kotlinpoet#1212] [Q&A]
I'm looking for a way to extract a proper `ClassName` from either a `TypeElement` or `KmClass` with a help of `kotlinpoet-metadata`. Specifically, I need a `ClassName` to generate a peer class for an existing class.

Prior to version 1.10, I've used to call  `ClassInspectorUtil.createClassName(kmClass.name)` for that. However, `ClassInspectorUtil` is internal in 1.10 and I'm probably missing some obvious why how it's exposed to the public code.

**Comment by ZacSweers** [comment]
That class was always in an internal package and never public API. For your use case, it's hard to advise without a concrete example as your question is a little vague

**Reply by GreyTeardrop** [reply]
Thanks, @ZacSweers!

To be more specific, I'm building an annotation processor, and using `TypeElement.toKmClass()` from the `kotlinpoet-metadata` to introspect types that the annotation processor receives. Ideally, I'd also like to get an instance of `com.squareup.kotlinpoet.ClassName` from that `TypeElement`, something like the `TypeElement.asClassName()` extension from the core `kotlinpoet` module, except that extension is marked as delicate and suggests using `kotlinpoet-metadata` interfaces instead.

Specifically, I use the `ClassName` extracted from the `TypeElement` to:
* create a new `ClassName` of a "builder" class for that `TypeElement` via `ClassName.peerClass()`
* reference fields from the original class in the builder's KDocs with the `ClassName.member()` extension

**Reply by ZacSweers** [reply]
Given that you have the package name and class name available to you, what's preventing you from constructing a `ClassName` with that information?

**Reply by GreyTeardrop** [reply]
That should work in most cases, thanks! I wonder if there could be some corner cases though (nested classes perhaps?). Would building `ClassName` from package name and simple name be better (cover more corner cases) than calling `TypeElement.asClassName()`?

**Reply by ZacSweers** [reply]
You can also just use it if you're worried about that

**Reply by GreyTeardrop** [reply]
Gotcha.

Should I ignore the warning on the `TypeElement.asClassName()` method ("Element APIs don't give complete information on Kotlin types. Consider using the kotlinpoet-metadata APIs instead.")?

As another alternative, could a PR that adds a new extension to the `kotlinpoet-metadata`, something like `TypeElement.asClassName()` from the core `kotlinpoet` module, but delegating to the `ClassInspectorUtil.createClassName` under the hood, be considered? I'd like to work on one if that is an acceptable approach?
